### PR TITLE
RBAC bucket user support for SG tests

### PR DESCRIPTION
**Pull request by raghusarangapani**
#### Fixes #.
https://github.com/couchbaselabs/mobile-testkit/issues/1103

- [x] Ran `flake8`
- [x] Ran `run_repo_tests.sh`

#### Changes proposed in this pull request:
- Create a user - username=bucketname with cluster_admin and bucket_admin rights
- Modify SG configs to use bucket auth

### CODE REVIEW

File: resources/sync_gateway_configs/bucket_online_offline/bucket_online_offline_default_di.json [hunk 1]
```diff
@@ -10,16 +10,20 @@
...
     "databases":{
         "db":{
             "feed_type":"DCPSHARD",
             "server":"http://{{ couchbase_server_primary_node }}:8091",
             "bucket":"data-bucket",
+            "password": "password",
             "channel_index":{
                 "server":"http://{{ couchbase_server_primary_node }}:8091",
                 "bucket":"index-bucket",
+                "username":"index-bucket",
```
Comment (sethrosetter) [inline resources/sync_gateway_configs/bucket_online_offline/bucket_online_offline_default_di.json]: Should we file an inconsistency bug in behavior between data / index buckets. Sometimes a username is required and sometime it is not. This is a bit confusing.

Comment (raghusarangapani) [inline resources/sync_gateway_configs/bucket_online_offline/bucket_online_offline_default_di.json]: Already exists - https://github.com/couchbase/sync_gateway/issues/2341

File: keywords/couchbaseserver.py [hunk 2]
```diff
@@ -58,6 +58,30 @@ def verify_server_version(host, expected_server_version):
...
+    }
+
+    log_info("Creating RBAC user {} with password {} and roles {}".format(bucketname, password, roles))
+
+    rbac_url = "{}/settings/rbac/users/builtin/{}".format(url, bucketname)
+
+    resp = ""
+    try:
+        resp = requests.put(rbac_url, data=data_user_params, auth=('Administrator', 'password'))
+    except:
```
Comment (raghusarangapani) [inline keywords/couchbaseserver.py]: Done

File: keywords/couchbaseserver.py [hunk 3]
```diff
@@ -265,18 +289,41 @@ def create_bucket(self, name, ram_quota_mb=1024):
...
-            "proxyPort": "11211",
-            "bucketType": "couchbase",
-            "flushEnabled": "1"
-        }
+        server_version = get_server_version(self.host)
+        server_major_version = int(server_version.split(".")[0])
+
+        if server_major_version <= 4:
+            # Create a bucket with password for server_major_version < 5
+            data = {
```
Comment (raghusarangapani) [inline keywords/couchbaseserver.py]: Done

File: keywords/couchbaseserver.py [hunk 4]
```diff
@@ -58,6 +58,30 @@ def verify_server_version(host, expected_server_version):
...
+        "password": password
+    }
+
+    log_info("Creating RBAC user {} with password {} and roles {}".format(bucketname, password, roles))
+
+    rbac_url = "{}/settings/rbac/users/builtin/{}".format(url, bucketname)
+
+    resp = ""
+    try:
+        resp = requests.put(rbac_url, data=data_user_params, auth=('Administrator', 'password'))
```
Comment (sethrosetter) [inline keywords/couchbaseserver.py]: I think we need a resp.raise_for_status()

File: libraries/testkit/server.py [hunk 5]
```diff
@@ -86,6 +87,13 @@ def delete_bucket(self, name):
         return status
 
     def create_buckets(self, names):
+        # Create a user with username=bucketname
+        if type(names) is list:
```
Comment (sethrosetter) [inline libraries/testkit/server.py]: Only create RBAC user if they do not exist

File: keywords/couchbaseserver.py [hunk 6]
```diff
@@ -58,6 +58,30 @@ def verify_server_version(host, expected_server_version):
         raise ProvisioningError("Unsupported version format")
 
 
+def create_internal_rbac_bucket_user(url, bucketname):
+    # Create user with username=bucketname and assign role
+    # bucket_admin and cluster_admin
+    roles = "cluster_admin,bucket_admin[{}]".format(bucketname)
```
Comment (sethrosetter) [inline keywords/couchbaseserver.py]: Check if user exists, if so, return

File: keywords/couchbaseserver.py (hunk 4 above)
Comment (raghusarangapani) [inline keywords/couchbaseserver.py]: Done

File: keywords/couchbaseserver.py (hunk 6 above)
Comment (raghusarangapani) [inline keywords/couchbaseserver.py]: I added a function to delete users and it gets called when bucket gets deleted. We can check if the user exists but not if the password is what we are expecting

File: libraries/testkit/server.py (hunk 5 above)
Comment (raghusarangapani) [inline libraries/testkit/server.py]: I added a function to delete users and it gets called when bucket gets deleted. We can check if the user exists but not if the password is what we are expecting

**Comment by raghusarangapani** [comment]
5.0.0 run with new changes:
http://uberjenkins.sc.couchbase.com:8080/job/cen7-sync-gateway-functional-tests-topology-specific-di/311/console

**Comment by sethrosetter** [comment]
Looks good!

**Pull request merged**